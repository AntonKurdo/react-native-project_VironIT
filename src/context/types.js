export const CHANGE_THEME = 'CHANGE_THEME';
export const SET_LOCATION = 'SET_LOCATION';
export const SET_MAP_MODAL = 'SET_MAP_MODAL';
export const SHOW_LOADER = 'SHOW_LOADER';
export const HIDE_LOADER = 'HIDE_LOADER';