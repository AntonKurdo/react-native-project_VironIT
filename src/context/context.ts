import { createContext, useContext } from 'react';
import { iNewShop } from '../screens/CreateShopForm.screen';
import { LatLng } from 'react-native-maps';

type iContext = {
  state?: iState,
  coords?: LatLng,     
  changeTheme?: () => void,
  setLocation?:  (location: string, coords: LatLng ) => void,
  setMapModal?: () => void,
  showLoader?: () => void,
  hideLoader?: () => void,
  setActiveUser?: (useName: string) => void, 
  clearActiveUser?: () => void,
  setShops?:  (shop: iNewShop) => void,
  clearShops?:  () => void,
  setRadius?:  (radius: number) => void      
};

interface iState {
  activeUser: string,
  isLightenMode: boolean,
  location: null | string,
  coords: LatLng,
  isMapVisible: boolean,
  isLoader: boolean,
  shops: Array<iNewShop>,       
  radius: number
};


export const useAppContext = () => useContext(AppContext); 
const AppContext =  createContext<iContext>({});

export default AppContext;