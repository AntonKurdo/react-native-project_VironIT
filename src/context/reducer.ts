import { CHANGE_THEME, SET_LOCATION, SET_MAP_MODAL, SHOW_LOADER, HIDE_LOADER, SET_ACTIVE_USER, CLEAR_ACTIVE_USER, SET_SHOPS, CLEAR_SHOPS } from './types';

interface iCoords {
  latitude: number | null, 
  longitude: number | null
}

interface iAction {
  type: string, 
  location?: string,
  coords?: iCoords,
  userName?: string,
  shops?: object[]
};

interface iState {  
  isLightenMode: boolean,
  location: null | string,
  coords: iCoords
};

const handlers = {
  [CHANGE_THEME]: (state: iState) => ({...state, isLightenMode: !state.isLightenMode}),
  [SET_LOCATION]: (state: iState, {location, coords}: iAction) => ({...state, location, coords}),
  [SET_MAP_MODAL]: state => ({...state, isMapVisible: !state.isMapVisible}),
  [SHOW_LOADER]: state => ({...state, isLoader: true}),
  [HIDE_LOADER]: state => ({...state, isLoader: false}),
  [SET_ACTIVE_USER]: (state, {userName}) => ({...state, activeUser: userName}),
  [CLEAR_ACTIVE_USER]: (state) => ({...state, activeUser: ''}),
  [SET_SHOPS]: (state, {shops}) => ({...state, shops}),
  [CLEAR_SHOPS]: state => ({...state, shops: []})
};


export const reducer = (state: iState, action: iAction): any => { 
  return handlers[action.type](state, action);
};