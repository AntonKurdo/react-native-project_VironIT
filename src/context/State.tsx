import React, {useReducer} from 'react';
import {Context} from './context';
import {reducer} from './reducer';
import { CHANGE_THEME, SET_LOCATION, SET_MAP_MODAL, SHOW_LOADER, HIDE_LOADER, SET_ACTIVE_USER, CLEAR_ACTIVE_USER, SET_SHOPS, CLEAR_SHOPS, SET_RADIUS } from './types';

export const AppState = ({children}) => {

    const INITIAL_STATE = {   
        activeUser: '',
        isLightenMode: true,
        location: null,
        coords: {
            latitude: null, 
            longitude: null
        },
        isMapVisible: false,
        isLoader: false,
        shops: [],       
        radius: '0'
    };

    const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
 
    const changeTheme : any = () => dispatch({type: CHANGE_THEME})
    const setLocation : any = (location, coords) => dispatch({type: SET_LOCATION, location, coords});
    const setMapModal : any = () => dispatch({type: SET_MAP_MODAL});
    const showLoader : any = () => dispatch({type: SHOW_LOADER});
    const hideLoader : any = () => dispatch({type: HIDE_LOADER});
    const setActiveUser : any = (userName) => dispatch({type: SET_ACTIVE_USER, userName});
    const clearActiveUser : any = () => dispatch({type: CLEAR_ACTIVE_USER});
    const setShops : any = (shops) => dispatch({type: SET_SHOPS, shops});
    const clearShops : any = () => dispatch({type: CLEAR_SHOPS}); 
    const setRadius : any = (radius) => dispatch({type: SET_RADIUS, radius}); 
 

    return <Context.Provider
        value={{
        state,
        coords: state.coords,     
        changeTheme,
        setLocation,
        setMapModal,
        showLoader,
        hideLoader,
        setActiveUser, 
        clearActiveUser,
        setShops,
        clearShops,
        setRadius       
    }}>{children}</Context.Provider>
};