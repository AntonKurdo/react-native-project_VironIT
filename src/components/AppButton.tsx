import React, {FC, useContext} from 'react';
import {StyleSheet, Text, TouchableOpacity} from 'react-native';
import { FontAwesome } from '@expo/vector-icons';
import { THEME } from './../theme';
import { Context } from './../context/context';


interface StylesType {
  [key: string]: string | number
};

type VariantType = 'V1' | 'V2';

type AppButtonProps = {
    children: any, 
    onPress: () => void,     
    iconName: any,
    iconSize?: number, 
    style?: StylesType,
    variant?: VariantType
};

export const AppButton : FC <AppButtonProps> = ({onPress, children, iconName, iconSize = 24,  style = {}, variant = 'V1'}) => {
    const {state} = useContext(Context);

    return (
      <TouchableOpacity style={variant === 'V1' ? state.isLightenMode ? {...styles.button, ...styles.buttonLightV1, ...style} : {...styles.button, ...styles.buttonDarkV1, ...style} : state.isLightenMode ? {...styles.button, ...styles.buttonLightV2, ...style} : {...styles.button, ...styles.buttonDarkV2, ...style}} onPress={onPress}>              
         <FontAwesome name={iconName} size={iconSize} color={variant === 'V1' ? state.isLightenMode ? THEME.MAIN_COLOR_LIGHT : THEME.MAIN_COLOR_DARK : state.isLightenMode ? THEME.SECOND_COLOR_LIGHT : THEME.SECOND_COLOR_LIGHT } />
         <Text style={variant === 'V1' ? state.isLightenMode ? {...styles.text, ...styles.textLightV1} : {...styles.text, ...styles.textDarkV1} : state.isLightenMode ? {...styles.text, ...styles.textLightV2} : {...styles.text, ...styles.textDarkV2}}>{children}</Text>
      </TouchableOpacity>        
    )
};

const styles = StyleSheet.create({
  button: {  
    flexDirection: 'row',
    width: THEME.WIDTH / 3,
    height: 50,
    justifyContent: 'center',
    paddingLeft: 5,
    alignItems: 'center',
    borderWidth: 2,   
    borderRadius:  THEME.BORDER_RADIUS
  },
  buttonLightV1: {
    borderColor: THEME.MAIN_COLOR_LIGHT,
  },
  buttonDarkV1: {
    borderColor: THEME.MAIN_COLOR_DARK,   
  },
  buttonLightV2: {
    borderColor: THEME.SECOND_COLOR_LIGHT,
  },
  buttonDarkV2: {
    borderColor: THEME.SECOND_COLOR_LIGHT,   
  },
  text: {   
    marginLeft: 7,
    fontSize: 22,    
    fontWeight: 'bold'
  },
  textLightV1: {
    color: THEME.MAIN_COLOR_LIGHT,
  },
  textDarkV1: {
    color: THEME.MAIN_COLOR_DARK,  
  },
  textLightV2: {
    color: THEME.SECOND_COLOR_LIGHT,
  },
  textDarkV2: {
    color: THEME.SECOND_COLOR_LIGHT,  
  }
})