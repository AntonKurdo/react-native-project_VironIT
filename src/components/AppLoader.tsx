import React, { FC } from 'react';
import {StyleSheet, View, ActivityIndicator} from 'react-native';
import { THEME } from './../theme';
import { useAppContext } from './../context/context';

export const AppLoader: FC = () => {
  const {state} = useAppContext();
  
  return (
    <View style={styles.container}>   
      <ActivityIndicator size='large' color={state.isLightenMode ? THEME.MAIN_COLOR_LIGHT : THEME.MAIN_COLOR_DARK} /> 
    </View>    
  )
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  }
});

