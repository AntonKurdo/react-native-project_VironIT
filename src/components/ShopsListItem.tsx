import React, {FC, useContext} from 'react';
import {StyleSheet, View, TouchableOpacity, Text, Alert} from 'react-native';
// import { iNewShop } from '../screens/CreateShopForm.screen';
import { THEME } from './../theme';
import { AntDesign } from '@expo/vector-icons';
import { Context } from './../context/context';
import { getData } from './../services/asyncStorage.service';
import AsyncStorage from '@react-native-async-storage/async-storage';


interface ShopsListItemProps {
  // item: iNewShop  
  item: any
};

export const ShopsListItem: FC<ShopsListItemProps> = ({item}) => {

  const {state, setShops} = useContext(Context); 

  const likeShop = async (id) => {
    try{
      const data = JSON.parse(await getData());
      const user = data.find(user => user.login === state.activeUser);
      const newShops = user.shops.map(shop => {
        if(shop.id === id) {
          shop.isFavourite = !shop.isFavourite
        }
        return shop;
      });
      setShops(newShops);
      await AsyncStorage.setItem('users', JSON.stringify(data));     
    } catch(e) {
      console.log(e)
    }
  };

  const removeShop = async (id: string, name: string): Promise<any> => {
    Alert.alert(
      "Removing shop...",
      `Are sure to remove ${name}?`,
      [
        {
          text: "Cancel",          
          style: "cancel"
        },
        { text: "OK", onPress: async () => {
          try{
            const data = JSON.parse(await getData());
            const user = data.find(user => user.login === state.activeUser);
            const newShops = user.shops.filter(shop => shop.id !== id);
            setShops(newShops);
            user.shops = newShops;    
            await AsyncStorage.setItem('users', JSON.stringify(data));           
          } catch(e) {
            console.log(e)
          }
        }}
      ],
      { cancelable: false }
    );   
  };



  return (
    <TouchableOpacity  
      style={state.isLightenMode ? {...styles.container, ...styles.containerLight} :  {...styles.container, ...styles.containerDark}} 
      onPress={likeShop.bind(null, item.id)}
      onLongPress={removeShop.bind(null, item.id, item.name)}
    >    
        <Text style={state.isLightenMode ? {...styles.text, ...styles.textLight} : {...styles.text, ...styles.textDark}}>{item.name}</Text>
        <Text style={styles.subtext}>Lat {item.latitude}</Text>
        <Text style={styles.subtext}>Long {item.longitude}</Text>
        {
          item.isFavourite && 
          <View style={styles.btnLike}>
            <AntDesign name="like1" size={28} color={state.isLightenMode ? THEME.MAIN_COLOR_LIGHT : THEME.MAIN_COLOR_DARK} />
          </View>
        }        
    </TouchableOpacity>
  )
};


const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    borderWidth: 2,
    borderColor: 'black',
    width: THEME.WIDTH * 0.8,
    borderRadius: 5,
    paddingVertical: 5,
    marginBottom: 10    
  },
  containerLight: {
    borderColor: THEME.MAIN_COLOR_LIGHT,
  }, 
  containerDark: {
    borderColor: THEME.MAIN_COLOR_DARK,
  }, 
  text: {
    fontSize: 20,
    fontWeight: '700'    
  },
  textLight: {
    color: THEME.MAIN_COLOR_LIGHT
  },
  textDark: {
    color: THEME.MAIN_COLOR_DARK
  },
  subtext: {
    fontSize: 10,
    color: 'gray'
  },
  btnLike: {
    position: 'absolute',
    right: 10,
    top: '30%'
  }
})