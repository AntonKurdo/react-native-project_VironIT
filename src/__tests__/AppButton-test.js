import React from 'react';
import {configure, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { AppButton } from './../components/AppButton';
import * as AppContext from '../context/context';

configure({adapter: new Adapter()});

describe('<AppButton />' , () => {
  test('AppButton render correctly', () => {    
  const contextValue = {state: {isLightenMode: false}};
  jest
    .spyOn(AppContext, 'useAppContext')
    .mockImplementation(() => contextValue);
  const wrapper = shallow(<AppButton />);
  expect(wrapper).toMatchSnapshot();
  });
});