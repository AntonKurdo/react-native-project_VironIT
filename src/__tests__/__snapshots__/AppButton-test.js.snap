// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppButton /> AppButton render correctly 1`] = `ShallowWrapper {}`;
